File: fixtures/VbFixture/AsyncPatterns.vb
Imports System.Threading.Tasks
Imports System.Collections.Generic

Namespace Fixtures

    ' Covers: interface with Async function signatures
    Public Interface IDataFetcher
        Function FetchAsync(url As String) As Task(Of String)
        Function FetchManyAsync(urls As IEnumerable(Of String)) As Task(Of String())
    End Interface

    ' Covers: Async Function returning Task(Of T), Await expression,
    '         Async Function returning Task(Of T()), method-to-method call
    Public Class HttpDataFetcher
        Implements IDataFetcher

        Public Async Function FetchAsync(url As String) As Task(Of String) _
            Implements IDataFetcher.FetchAsync
            Await Task.Delay(1)
            Return String.Format("response from {0}", url)
        End Function

        Public Async Function FetchManyAsync(urls As IEnumerable(Of String)) As Task(Of String()) _
            Implements IDataFetcher.FetchManyAsync
            Dim results = New List(Of String)()
            For Each url In urls
                Dim result = Await FetchAsync(url)
                results.Add(result)
            Next
            Return results.ToArray()
        End Function

    End Class

    ' Covers: Async Function returning Task(Of Integer), Await on interface method,
    '         Async Sub (fire-and-forget), non-async method calling async
    Public Class UrlProcessor
        Implements IDataFetcher

        Private ReadOnly _inner As IDataFetcher

        Public Sub New(inner As IDataFetcher)
            _inner = inner
        End Sub

        Public Async Function FetchAsync(url As String) As Task(Of String) _
            Implements IDataFetcher.FetchAsync
            Return Await _inner.FetchAsync(url)
        End Function

        Public Async Function FetchManyAsync(urls As IEnumerable(Of String)) As Task(Of String()) _
            Implements IDataFetcher.FetchManyAsync
            Return Await _inner.FetchManyAsync(urls)
        End Function

        Public Async Function ProcessUrlAsync(url As String) As Task(Of Integer)
            Dim content = Await FetchAsync(url)
            Return content.Length
        End Function

        ' Covers: Async Sub (isAsync = true, return type is void-equivalent)
        Public Async Sub FireAndForget(url As String)
            Await ProcessUrlAsync(url)
        End Sub

        Public Function GetLengthSync(url As String) As Integer
            Return ProcessUrlAsync(url).GetAwaiter().GetResult()
        End Function

    End Class

End Namespace

File: fixtures/CsFixture/AsyncPatterns.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Linq;

namespace Fixtures
{
    // Covers: interface with async method signatures
    public interface IDataFetcher
    {
        Task<string>   FetchAsync(string url);
        Task<string[]> FetchManyAsync(IEnumerable<string> urls);
    }

    // Covers: async Task<T> implementation, await expression, method-to-method call
    public class HttpDataFetcher : IDataFetcher
    {
        public async Task<string> FetchAsync(string url)
        {
            await Task.Delay(1);
            return $"response from {url}";
        }

        public async Task<string[]> FetchManyAsync(IEnumerable<string> urls)
        {
            var tasks = urls.Select(u => FetchAsync(u));
            return await Task.WhenAll(tasks);
        }
    }

    // Covers: async Task<int>, await on injected interface, async void (fire-and-forget),
    //         sync method calling async via .GetAwaiter().GetResult()
    public class UrlProcessor : IDataFetcher
    {
        private readonly IDataFetcher _inner;

        public UrlProcessor(IDataFetcher inner)
        {
            _inner = inner;
        }

        public async Task<string> FetchAsync(string url) =>
            await _inner.FetchAsync(url);

        public async Task<string[]> FetchManyAsync(IEnumerable<string> urls) =>
            await _inner.FetchManyAsync(urls);

        public async Task<int> ProcessUrlAsync(string url)
        {
            var content = await FetchAsync(url);
            return content.Length;
        }

        // Covers: async void method (isAsync = true)
        public async void FireAndForget(string url)
        {
            await ProcessUrlAsync(url);
        }

        public int GetLengthSync(string url) =>
            ProcessUrlAsync(url).GetAwaiter().GetResult();
    }
}
